File: core/src/main/java/net/jselby/escapists/game/ObjectInstance.java
package net.jselby.escapists.game;

import net.jselby.escapists.data.ObjectDefinition;
import net.jselby.escapists.data.chunks.ObjectInstances;
import net.jselby.escapists.data.objects.ObjectCommon;
import org.mini2Dx.core.graphics.Graphics;

/**
 * A instance of an object within the world.
 *
 * @author j_selby
 */
public abstract class ObjectInstance {
    private final String name;
    private final int id;
    private final int layerId;
    private final int objectInfo;

    private int x;
    private int y;

    private boolean isVisible = true;

    public ObjectInstance(ObjectDefinition definition, ObjectInstances.ObjectInstance instance) {
        this.name = definition.name;
        this.id = instance.getHandle();
        this.objectInfo = instance.getObjectInfo();
        this.layerId = instance.getLayer();
        this.x = instance.getX();
        this.y = instance.getY();

        if (definition.properties.isCommon()) {
            isVisible = ((ObjectCommon) definition.properties.getProperties()).isVisibleAtStart();
        }
    }

    /**
     * Returns the ID of this object.
     * @return An object ID
     */
    public int getID() {
        return id;
    }

    /**
     * Returns the name of this object.
     * @return A String value
     */
    public String getName() {
        return name;
    }

    /**
     * Returns the layer that this object exists on. Should not change post init.
     * @return The layer ID of this object.
     */
    public int getLayerID() {
        return layerId;
    }

    public float getX() {
        return x;
    }

    public float getY() {
        return y;
    }

    public void setX(int newX) {
        x = newX;
    }

    public void setY(int newY) {
        y = newY;
    }

    /**
     * Returns the width of this object
     * @return A width argument
     */
    public abstract float getWidth();

    /**
     * Returns the height of this object
     * @return A height argument
     */
    public abstract float getHeight();

    /**
     * Ticks this object, accepting input and responding accordingly.
     * @param container The container to poll information from.
     */
    public abstract void tick(EscapistsGame container);

    /**
     * Draws this object onto the screen.
     * @param container The container to poll information from.
     * @param g The graphics instance to draw stuff onto.
     */
    public abstract void draw(EscapistsGame container, Graphics g);

    public boolean isVisible() {
        return isVisible;
    }

    public void setIsVisible(boolean isVisible) {
        this.isVisible = isVisible;
    }

    /**
     * Returns this Object's ObjectInfo information.
     * @return A ObjectInfo handle.
     */
    public int getObjectInfo() {
        return objectInfo;
    }

}

File: core/src/main/java/net/jselby/escapists/game/ObjectInstance.kt
package net.jselby.escapists.game

import net.jselby.escapists.data.ObjectDefinition
import net.jselby.escapists.data.chunks.ObjectInstances
import net.jselby.escapists.data.objects.ObjectCommon
import org.mini2Dx.core.graphics.Graphics
import java.util.*

/**
 * A instance of an object within the world.

 * @author j_selby
 */
abstract class ObjectInstance(definition: ObjectDefinition, instance: ObjectInstances.ObjectInstance) {
    /**
     * Returns the name of this object.
     * @return A String value
     */
    val name: String
    /**
     * Returns the ID of this object.
     * @return An object ID
     */
    private val id: Int
    /**
     * Returns the layer that this object exists on. Should not change post init.
     * @return The layer ID of this object.
     */
    val layerID: Int
    /**
     * Returns this Object's ObjectInfo information.
     * @return A ObjectInfo handle.
     */
    val objectInfo: Int

    open var x: Float = 0f
    open var y: Float = 0f

    var animation = 0
    var animationFrame = 0
    var imageAlpha = 255

    var bold = false

    var isVisible = true

    val namedVariables: Map<String, Any> = HashMap()

    val configVariables: Map<String, Any> = HashMap()

    val alterableValues: IntArray = IntArray(26);
    val alterableStrings: Array<String> = Array(26, {i -> ""});

    var httpContentLoaded = false
    var httpContent = ""


    val listElements: List<String> = ArrayList()
    var selectedLine = 0
    var loadedFile = ""

    init {
        this.name = definition.name.trim()
        this.id = instance.handle
        this.objectInfo = instance.objectInfo
        this.layerID = instance.layer.toInt()
        this.x = instance.x.toFloat()
        this.y = instance.y.toFloat()

        if (definition.properties.isCommon) {
            val objectCommon = (definition.properties.properties as ObjectCommon?)!!;
            isVisible = objectCommon.isVisibleAtStart;

            if (objectCommon.values != null) {
                var i = 0;
                for (value in objectCommon.values.values) {
                    alterableValues[i++] = value;
                }
            }

            if (objectCommon.strings != null) {
                var i = 0;
                for (value in objectCommon.strings.strings) {
                    alterableStrings[i++] = value;
                }
            }
        }
    }

    /**
     * Returns the width of this object
     * @return A width argument
     */
    abstract val width: Float

    /**
     * Returns the height of this object
     * @return A height argument
     */
    abstract val height: Float

    /**
     * Gets the X position of the object as it would be, rendered to the screen.
     * This is independent of cameras, and is simply accounting for centering, etc.
     */
    open fun getScreenX(): Float = x

    /**
     * Gets the Y position of the object as it would be, rendered to the screen.
     * This is independent of cameras, and is simply accounting for centering, etc.
     */
    open fun getScreenY(): Float = y

    /**
     * Ticks this object, accepting input and responding accordingly.
     * @param container The container to poll information from.
     */
    abstract fun tick(container: EscapistsGame)

    /**
     * Draws this object onto the screen.
     * @param container The container to poll information from.
     * *
     * @param g The graphics instance to draw stuff onto.
     */
    abstract fun draw(container: EscapistsGame, g: Graphics)
}
